### PR TITLE
[DI] Fix inlining conflict by restricting IteratorArgument to Reference[]

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

`Reference` found in `ArgumentInterface::getValue()` are currently not inlined.
While trying to do so (hint: I failed), I noticed that the current code is broken for `IteratorArgument` which can contain anonymous `Definition` for now, which are then not inlined correctly.

This PR restricts `IteratorArgument` to arrays of `Reference`, and improves a few related things found while doing it.

(fabbot failure is false positive)

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -683,16 +684,19 @@ private function addService($id, Definition $definition)
             $code .= sprintf("        @trigger_error(%s, E_USER_DEPRECATED);\n\n", $this->export($definition->getDeprecationMessage($id)));
         }
 
+        $inlinedDefinitions = $this->getInlinedDefinitions($definition);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: this replaces *many* calls to `getInlinedDefinitions` by a single one

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 2]
```diff
@@ -1332,8 +1336,6 @@ private function getDefinitionsFromArguments(array $arguments)
         foreach ($arguments as $argument) {
             if (is_array($argument)) {
                 $definitions = array_merge($definitions, $this->getDefinitionsFromArguments($argument));
-            } elseif ($argument instanceof ArgumentInterface) {
-                $definitions = array_merge($definitions, $this->getDefinitionsFromArguments($argument->getValues()));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: this is a bug, silent for now because we don't have inline definitions in `$argument->getValues()`, but still a logical mistake

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 3]
```diff
@@ -1452,6 +1454,9 @@ private function dumpValue($value, $interpolate = true)
             if ($value->getMethodCalls()) {
                 throw new RuntimeException('Cannot dump definitions which have method calls.');
             }
+            if ($value->getProperties()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: should be done on 2.7 also

Comment (stof) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: then please split it in a separate PR

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 4]
```diff
@@ -654,14 +654,18 @@ private function resolveServices($value, $file, $isParameter = false)
             $argument = $value->getValue();
             if ('iterator' === $value->getTag()) {
                 if (!is_array($argument)) {
-                    throw new InvalidArgumentException('"!iterator" tag only accepts sequences.');
+                    throw new InvalidArgumentException(sprintf('"!iterator" tag only accepts sequences in "%s".', $file));
                 }
 
-                return new IteratorArgument($this->resolveServices($argument, $file, $isParameter));
+                try {
+                    return new IteratorArgument($this->resolveServices($argument, $file, $isParameter));
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: IMO, `resolveServices` should be outside the try/catch, so that we don't replace its exceptions by mistake

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: done in #22498

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: good idea, done (also for the Xml loader, just in case)

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php [hunk 5]
```diff
@@ -480,7 +480,12 @@ private function getArgumentsAsPhp(\DOMElement $node, $name, $lowercase = true,
                     $arguments[$key] = $this->getArgumentsAsPhp($arg, $name, false);
                     break;
                 case 'iterator':
-                    $arguments[$key] = new IteratorArgument($this->getArgumentsAsPhp($arg, $name, false));
+                    $arg = $this->getArgumentsAsPhp($arg, $name, false);
+                    try {
+                        $arguments[$key] = new IteratorArgument($arg);
+                    } catch (InvalidArgumentException $e) {
+                        throw new InvalidArgumentException(sprintf('Tag "<%s>" with type="iterator" only accepts collections of type="service" references.', $name));
```
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: `Tag` => `Argument`?

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php (hunk 5 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: its an xml name, eg `Tag "<argument>" ...` or `Tag "<property>" ...`

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.
